### ISSUE TITLE
Rule proposal: `no-set-duplicates`

### ISSUE THREAD

**Issue by futpib** [sindresorhus/eslint-plugin-unicorn#1559]
Sometimes a hardcoded `Set` grows large enough for it to be hard to see if there is already a value in it. One could miss that a value one wants to add is already there and add a duplicate.

This rule should at minimum report hardcoded sets of literals and `const`s with duplicates.

This rule may also report consecutive `.add` calls with same arguments (this is probably much harder to implement). Or maybe this could be a separate rule like `unicorn/no-array-push-push`.

## Fail

```js
const allowed = new Set([
	'foo',
	'bar',
	'buz',
	'qux',
	'bar', // ← Remove duplicate value `'bar'` from the Set
]);
```

## Pass

```js
const allowed = new Set([
	'foo',
	'bar',
	'buz',
	'qux',
]);
```

**Comment by sindresorhus** [comment]
How about naming it `no-duplicate-literal-elements`?

**Comment by sindresorhus** [comment]
@quirimmo This is already answered in the issue body:

> Sometimes a hardcoded Set grows large enough for it to be hard to see if there is already a value in it. One could miss that a value one wants to add is already there and add a duplicate.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by SidStraw** [comment]
I am looking at this issue.

**Comment by SidStraw** [comment]
I have added a rule about `Set` and `Map`.

PTAL, thx.

**Comment by SidStraw** [comment]
Hello @futpib @sindresorhus 

There are two questions here that I need to ask you to confirm.

https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1635#discussion_r771135771
https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1635#issuecomment-996478812

Thx~

**Comment by SidStraw** [comment]
Actually, I’m not sure why I need to check the arrays.

There are many cases where the contents of the array are duplicated.
like this: ((x, y) => Math.pow(x, y)).apply(null, [2, 2])

https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1635#discussion_r772027700

I should understand exactly.
To know how to describe.

**Comment by fisker** [comment]
Let's focus on `Set`?

**Comment by fisker** [comment]
And let's try to statically analyze the AST too, not only literals.

## Fail

```js
new Set([-1, -1]);
```

```js
const foo = 2;
new Set([foo, 2]);
```

```js
new Set([foo.bar, foo.bar]);
```

**Comment by sindresorhus** [comment]
If anyone wants to work on this, see the initial attempt and feedback in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1635.
